In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
test= pd.read_csv("/content/sample_data/1.test_convertito_numerici.csv", sep=",")
test_non = pd.read_csv("/content/sample_data/1.test_non_convertito.csv", sep=",")
train_non = pd.read_csv("/content/sample_data/1.train_non_convertito.csv", sep=",")
train = pd.read_csv("/content/sample_data/1.train_convertito_numerici.csv", sep=",")

Quante righe e colonne sono presenti nei due dataset? I valori sono per lo più numerici o categorici e tra le righe sono poi presenti valori nulli?

In [2]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56707 entries, 0 to 56706
Data columns (total 30 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   hotel                           56707 non-null  int64  
 1   is_canceled                     56707 non-null  int64  
 2   lead_time                       56707 non-null  int64  
 3   arrival_date_year               56707 non-null  int64  
 4   arrival_date_month              56707 non-null  int64  
 5   arrival_date_week_number        56707 non-null  int64  
 6   arrival_date_day_of_month       56707 non-null  int64  
 7   stays_in_weekend_nights         56707 non-null  int64  
 8   stays_in_week_nights            56707 non-null  int64  
 9   adults                          56707 non-null  int64  
 10  children                        56707 non-null  float64
 11  babies                          56707 non-null  int64  
 12  meal                            

In [3]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40687 entries, 0 to 40686
Data columns (total 30 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   hotel                           40687 non-null  int64  
 1   is_canceled                     40687 non-null  int64  
 2   lead_time                       40687 non-null  int64  
 3   arrival_date_year               40687 non-null  int64  
 4   arrival_date_month              40687 non-null  int64  
 5   arrival_date_week_number        40687 non-null  int64  
 6   arrival_date_day_of_month       40687 non-null  int64  
 7   stays_in_weekend_nights         40687 non-null  int64  
 8   stays_in_week_nights            40687 non-null  int64  
 9   adults                          40687 non-null  int64  
 10  children                        40687 non-null  float64
 11  babies                          40687 non-null  int64  
 12  meal                            

In [4]:
test.isnull().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          0
babies                            0
meal                              0
country                           0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
reserved_room_type                0
assigned_room_type                0
booking_changes                   0
deposit_type                      0
agent                             0
company                           0
days_in_waiting_list              0
customer_type                     0
adr                         

In [5]:
train.isnull().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          0
babies                            0
meal                              0
country                           0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
reserved_room_type                0
assigned_room_type                0
booking_changes                   0
deposit_type                      0
agent                             0
company                           0
days_in_waiting_list              0
customer_type                     0
adr                         

Nel "Train dataset" sono presenti 56707 righe mentre nel "Test Dataset" 40687. Le colonne, invece, sono 30 per entrambi. I valori sono perlopiù categorici, infatti,  molti dei valori segnati come "int" in realtà sono la conversione in numero di "Si" e "No". In entrambi i dateset non sono presenti valori nulli

Come si distribuisce il valore relativo al numero di persone? Quante prenotazioni includono bambini e quanti ospiti sono, invece,  da soli?

In [20]:
train["persone"] = train["babies"] + train["children"] + train["adults"]
print(train["persone"].mean())
test["persone"] = train["babies"] + train["children"] + train["adults"]
print(test["persone"].mean())

1.9633378595235156
1.9511637623811044


In [7]:
train.groupby(train["babies"]).count()

hotel  is_canceled  lead_time  arrival_date_year  arrival_date_month  \
babies                                                                         
0       56261        56261      56261              56261               56261   
1         437          437        437                437                 437   
2           8            8          8                  8                   8   
10          1            1          1                  1                   1   

        arrival_date_week_number  arrival_date_day_of_month  \
babies                                                        
0                          56261                      56261   
1                            437                        437   
2                              8                          8   
10                             1                          1   

        stays_in_weekend_nights  stays_in_week_nights  adults  ...  \
babies                                                         ...   
0                         56261                 56261   56261  ...   
1                           437                   437     437  ...   
2                             8                     8       8  ...   
10                            1                     1       1  ...   

        booking_changes  deposit_type  agent  company  days_in_waiting_list  \
babies                                                                        
0                 56261         56261  56261    56261                 56261   
1                   437           437    437      437                   437   
2                     8             8      8        8                     8   
10                    1             1      1        1                     1   

        customer_type    adr  required_car_parking_spaces  \
babies                                                      
0               56261  56261                        56261   
1                 437    437                          437   
2                   8      8                            8   
10                  1      1                            1   

        total_of_special_requests  persone  
babies                                      
0                           56261    56261  
1                             437      437  
2                               8        8  
10                              1        1  

[4 rows x 30 columns]

In [8]:
#unisco colonne children e babies
train["colonna_bambini"] = train["babies"] + train["children"]
train.groupby(train["colonna_bambini"]).count()

hotel  is_canceled  lead_time  arrival_date_year  \
colonna_bambini                                                     
0.0              52211        52211      52211              52211   
1.0               2678         2678       2678               2678   
2.0               1768         1768       1768               1768   
3.0                 49           49         49                 49   
10.0                 1            1          1                  1   

                 arrival_date_month  arrival_date_week_number  \
colonna_bambini                                                 
0.0                           52211                     52211   
1.0                            2678                      2678   
2.0                            1768                      1768   
3.0                              49                        49   
10.0                              1                         1   

                 arrival_date_day_of_month  stays_in_weekend_nights  \
colonna_bambini                                                       
0.0                                  52211                    52211   
1.0                                   2678                     2678   
2.0                                   1768                     1768   
3.0                                     49                       49   
10.0                                     1                        1   

                 stays_in_week_nights  adults  ...  booking_changes  \
colonna_bambini                                ...                    
0.0                             52211   52211  ...            52211   
1.0                              2678    2678  ...             2678   
2.0                              1768    1768  ...             1768   
3.0                                49      49  ...               49   
10.0                                1       1  ...                1   

                 deposit_type  agent  company  days_in_waiting_list  \
colonna_bambini                                                       
0.0                     52211  52211    52211                 52211   
1.0                      2678   2678     2678                  2678   
2.0                      1768   1768     1768                  1768   
3.0                        49     49       49                    49   
10.0                        1      1        1                     1   

                 customer_type    adr  required_car_parking_spaces  \
colonna_bambini                                                      
0.0                      52211  52211                        52211   
1.0                       2678   2678                         2678   
2.0                       1768   1768                         1768   
3.0                         49     49                           49   
10.0                         1      1                            1   

                 total_of_special_requests  persone  
colonna_bambini                                      
0.0                                  52211    52211  
1.0                                   2678     2678  
2.0                                   1768     1768  
3.0                                     49       49  
10.0                                     1        1  

[5 rows x 31 columns]

In [23]:
test["colonna_bambini"] = test["babies"] + test["children"]
test.groupby(test["colonna_bambini"]).count()

hotel  is_canceled  lead_time  arrival_date_year  \
colonna_bambini                                                     
0.0              36941        36941      36941              36941   
1.0               2136         2136       2136               2136   
2.0               1565         1565       1565               1565   
3.0                 45           45         45                 45   

                 arrival_date_month  arrival_date_week_number  \
colonna_bambini                                                 
0.0                           36941                     36941   
1.0                            2136                      2136   
2.0                            1565                      1565   
3.0                              45                        45   

                 arrival_date_day_of_month  stays_in_weekend_nights  \
colonna_bambini                                                       
0.0                                  36941                    36941   
1.0                                   2136                     2136   
2.0                                   1565                     1565   
3.0                                     45                       45   

                 stays_in_week_nights  adults  ...  booking_changes  \
colonna_bambini                                ...                    
0.0                             36941   36941  ...            36941   
1.0                              2136    2136  ...             2136   
2.0                              1565    1565  ...             1565   
3.0                                45      45  ...               45   

                 deposit_type  agent  company  days_in_waiting_list  \
colonna_bambini                                                       
0.0                     36941  36941    36941                 36941   
1.0                      2136   2136     2136                  2136   
2.0                      1565   1565     1565                  1565   
3.0                        45     45       45                    45   

                 customer_type    adr  required_car_parking_spaces  \
colonna_bambini                                                      
0.0                      36941  36941                        36941   
1.0                       2136   2136                         2136   
2.0                       1565   1565                         1565   
3.0                         45     45                           45   

                 total_of_special_requests  persone  
colonna_bambini                                      
0.0                                  36941    36941  
1.0                                   2136     2136  
2.0                                   1565     1565  
3.0                                     45       45  

[4 rows x 31 columns]

In [9]:
#media cancellazioni in base al numero di bambini
train.groupby(train["colonna_bambini"])["is_canceled"].mean()

colonna_bambini
0.0     0.360020
1.0     0.305078
2.0     0.400452
3.0     0.306122
10.0    0.000000
Name: is_canceled, dtype: float64

In [29]:
test.groupby(test["colonna_bambini"])["is_canceled"].mean()

colonna_bambini
0.0    0.385859
1.0    0.352060
2.0    0.466454
3.0    0.200000
Name: is_canceled, dtype: float64

In [10]:
train[train["colonna_bambini"] > 9.00]

hotel  is_canceled  lead_time  arrival_date_year  arrival_date_month  \
18765      0            0         37               2016                   1   

       arrival_date_week_number  arrival_date_day_of_month  \
18765                         3                         12   

       stays_in_weekend_nights  stays_in_week_nights  adults  ...  \
18765                        0                     2       2  ...   

       deposit_type  agent  company  days_in_waiting_list  customer_type  \
18765             0    9.0     -1.0                     0              2   

         adr  required_car_parking_spaces  total_of_special_requests  persone  \
18765  84.45                            0                          1     12.0   

       colonna_bambini  
18765             10.0  

[1 rows x 32 columns]

In [19]:
train.groupby(train["persone"]).count()


hotel  is_canceled  lead_time  arrival_date_year  arrival_date_month  \
persone                                                                         
0.0         84           84         84                 84                  84   
1.0      10967        10967      10967              10967               10967   
2.0      38563        38563      38563              38563               38563   
3.0       5189         5189       5189               5189                5189   
4.0       1852         1852       1852               1852                1852   
5.0         51           51         51                 51                  51   
12.0         1            1          1                  1                   1   

         arrival_date_week_number  arrival_date_day_of_month  \
persone                                                        
0.0                            84                         84   
1.0                         10967                      10967   
2.0                         38563                      38563   
3.0                          5189                       5189   
4.0                          1852                       1852   
5.0                            51                         51   
12.0                            1                          1   

         stays_in_weekend_nights  stays_in_week_nights  adults  ...  \
persone                                                         ...   
0.0                           84                    84      84  ...   
1.0                        10967                 10967   10967  ...   
2.0                        38563                 38563   38563  ...   
3.0                         5189                  5189    5189  ...   
4.0                         1852                  1852    1852  ...   
5.0                           51                    51      51  ...   
12.0                           1                     1       1  ...   

         booking_changes  deposit_type  agent  company  days_in_waiting_list  \
persone                                                                        
0.0                   84            84     84       84                    84   
1.0                10967         10967  10967    10967                 10967   
2.0                38563         38563  38563    38563                 38563   
3.0                 5189          5189   5189     5189                  5189   
4.0                 1852          1852   1852     1852                  1852   
5.0                   51            51     51       51                    51   
12.0                   1             1      1        1                     1   

         customer_type    adr  required_car_parking_spaces  \
persone                                                      
0.0                 84     84                           84   
1.0              10967  10967                        10967   
2.0              38563  38563                        38563   
3.0               5189   5189                         5189   
4.0               1852   1852                         1852   
5.0                 51     51                           51   
12.0                 1      1                            1   

         total_of_special_requests  colonna_bambini  
persone                                              
0.0                             84               84  
1.0                          10967            10967  
2.0                          38563            38563  
3.0                           5189             5189  
4.0                           1852             1852  
5.0                             51               51  
12.0                             1                1  

[7 rows x 31 columns]

In [28]:
test.groupby(test["persone"]).count()

hotel  is_canceled  lead_time  arrival_date_year  arrival_date_month  \
persone                                                                         
0.0         27           27         27                 27                  27   
1.0       7911         7911       7911               7911                7911   
2.0      28241        28241      28241              28241               28241   
3.0       3081         3081       3081               3081                3081   
4.0       1391         1391       1391               1391                1391   
5.0         35           35         35                 35                  35   
12.0         1            1          1                  1                   1   

         arrival_date_week_number  arrival_date_day_of_month  \
persone                                                        
0.0                            27                         27   
1.0                          7911                       7911   
2.0                         28241                      28241   
3.0                          3081                       3081   
4.0                          1391                       1391   
5.0                            35                         35   
12.0                            1                          1   

         stays_in_weekend_nights  stays_in_week_nights  adults  ...  \
persone                                                         ...   
0.0                           27                    27      27  ...   
1.0                         7911                  7911    7911  ...   
2.0                        28241                 28241   28241  ...   
3.0                         3081                  3081    3081  ...   
4.0                         1391                  1391    1391  ...   
5.0                           35                    35      35  ...   
12.0                           1                     1       1  ...   

         booking_changes  deposit_type  agent  company  days_in_waiting_list  \
persone                                                                        
0.0                   27            27     27       27                    27   
1.0                 7911          7911   7911     7911                  7911   
2.0                28241         28241  28241    28241                 28241   
3.0                 3081          3081   3081     3081                  3081   
4.0                 1391          1391   1391     1391                  1391   
5.0                   35            35     35       35                    35   
12.0                   1             1      1        1                     1   

         customer_type    adr  required_car_parking_spaces  \
persone                                                      
0.0                 27     27                           27   
1.0               7911   7911                         7911   
2.0              28241  28241                        28241   
3.0               3081   3081                         3081   
4.0               1391   1391                         1391   
5.0                 35     35                           35   
12.0                 1      1                            1   

         total_of_special_requests  colonna_bambini  
persone                                              
0.0                             27               27  
1.0                           7911             7911  
2.0                          28241            28241  
3.0                           3081             3081  
4.0                           1391             1391  
5.0                             35               35  
12.0                             1                1  

[7 rows x 31 columns]

In [32]:
train.groupby(train["persone"])["is_canceled"].mean()

persone
0.0     0.154762
1.0     0.301176
2.0     0.380468
3.0     0.305454
4.0     0.405508
5.0     0.254902
12.0    0.000000
Name: is_canceled, dtype: float64

In [30]:
test.groupby(test["persone"])["is_canceled"].mean()

persone
0.0     0.296296
1.0     0.307926
2.0     0.409865
3.0     0.344369
4.0     0.471603
5.0     0.228571
12.0    1.000000
Name: is_canceled, dtype: float64

La media del numero di persone è di circa 2 persone in entrambi i dataset.
Il numero di prenotazioni composte da un numero qualsiasi di adulti con almeno un bambino ("babies" e/o "children") è relativamente basso in entrambi i dataset (poco più di 2000 casi). Il numero massimo di bambini è di 10 raggiunto però in un singolo caso. Escludendo quello il massimo di bambini a seguito è di 3. Le cancellazioni aumentano all'aumentare del numero di bambini in entrambi i dataset (le prenotazioni con almeno 3 bambini sono in numero così ridotto da non essere rilevanti). Escludendo suddetto caso estremo il numero massimo di persone per una singola camera è 5, anche il numero di queste non sembra avere particolare impatto sulle cancellazioni. In entrambi i dataset poi sono presenti valori anomali corrispondenti a 0 persone per camera (84 nel train e 27 nel test).
Il numero di viaggiatori "solitari" è invece copicuo e simile in entrambi i dataset.
Il numero di adulti per prenotazione varia da un minomo di 1 fino a un massimo di 4.



Come si distrubuisce il numero delle prenotazioni nei mesi dell'anno e la media delle cancellazioni?

In [12]:
train.groupby(train["arrival_date_month"]).count()

hotel  is_canceled  lead_time  arrival_date_year  \
arrival_date_month                                                     
1                    2248         2248       2248               2248   
2                    3891         3891       3891               3891   
3                    4824         4824       4824               4824   
4                    5428         5428       5428               5428   
5                    5478         5478       5478               5478   
6                    5292         5292       5292               5292   
7                    4572         4572       4572               4572   
8                    5063         5063       5063               5063   
9                    5394         5394       5394               5394   
10                   6203         6203       6203               6203   
11                   4454         4454       4454               4454   
12                   3860         3860       3860               3860   

                    arrival_date_week_number  arrival_date_day_of_month  \
arrival_date_month                                                        
1                                       2248                       2248   
2                                       3891                       3891   
3                                       4824                       4824   
4                                       5428                       5428   
5                                       5478                       5478   
6                                       5292                       5292   
7                                       4572                       4572   
8                                       5063                       5063   
9                                       5394                       5394   
10                                      6203                       6203   
11                                      4454                       4454   
12                                      3860                       3860   

                    stays_in_weekend_nights  stays_in_week_nights  adults  \
arrival_date_month                                                          
1                                      2248                  2248    2248   
2                                      3891                  3891    3891   
3                                      4824                  4824    4824   
4                                      5428                  5428    5428   
5                                      5478                  5478    5478   
6                                      5292                  5292    5292   
7                                      4572                  4572    4572   
8                                      5063                  5063    5063   
9                                      5394                  5394    5394   
10                                     6203                  6203    6203   
11                                     4454                  4454    4454   
12                                     3860                  3860    3860   

                    children  ...  deposit_type  agent  company  \
arrival_date_month            ...                                 
1                       2248  ...          2248   2248     2248   
2                       3891  ...          3891   3891     3891   
3                       4824  ...          4824   4824     4824   
4                       5428  ...          5428   5428     5428   
5                       5478  ...          5478   5478     5478   
6                       5292  ...          5292   5292     5292   
7                       4572  ...          4572   4572     4572   
8                       5063  ...          5063   5063     5063   
9                       5394  ...          5394   5394     5394   
10                      6203  ...          6203   6203     6203   
11                      4454  ...          4454   4454     4454   
12        

In [13]:
#media cancellazioni in base al mese di arrivo
train.groupby(train["arrival_date_month"])["is_canceled"].mean()

arrival_date_month
1     0.247776
2     0.343613
3     0.306177
4     0.379698
5     0.349580
6     0.396070
7     0.327865
8     0.360458
9     0.374861
10    0.405288
11    0.367310
12    0.362176
Name: is_canceled, dtype: float64

In [14]:
train.groupby(train["arrival_date_month"])["persone"].mean()

arrival_date_month
1     1.705961
2     1.854279
3     1.919776
4     1.929071
5     1.902154
6     1.908730
7     2.209318
8     2.291132
9     1.972562
10    1.928099
11    1.797261
12    2.001554
Name: persone, dtype: float64


Le famiglie più numerose sono concentrate nei mesi di luglio/agosto/settembre e dicembre.
La media delle cancellazioni ha un picco a ottobre mentre raggiunge il valore più basso a gennaio.

Come impatta il tipo di prenotazione (tramite tour operator o agenzia di viaggi).

Controllare numero di prenotazioni precedenti e cancellazioni precedenti (fino a qui niko)

Controllare la distribuzione in base al tipo di hotel selezionato

Controllare la distribuzione in base ai giorni in attesa

In [15]:
train[(train["days_in_waiting_list"] > 1.00) & (train["days_in_waiting_list"] < 5.00)]

hotel  is_canceled  lead_time  arrival_date_year  arrival_date_month  \
11355      1            0         26               2016                   5   
17026      1            0          4               2016                  11   
19457      0            1         44               2016                   2   
19458      0            1         44               2016                   2   
19459      0            1         44               2016                   2   
...      ...          ...        ...                ...                 ...   
39163      0            0         81               2016                   3   
39202      0            0         82               2016                   3   
39203      0            0         82               2016                   3   
39207      0            0         81               2016                   3   
43562      0            0         18               2016                   6   

       arrival_date_week_number  arrival_date_day_of_month  \
11355                        19                          7   
17026                        46                         12   
19457                         8                         17   
19458                         8                         17   
19459                         8                         17   
...                         ...                        ...   
39163                        13                         25   
39202                        13                         26   
39203                        13                         26   
39207                        13                         25   
43562                        24                          8   

       stays_in_weekend_nights  stays_in_week_nights  adults  ...  \
11355                        3                     6       2  ...   
17026                        1                     1       2  ...   
19457                        0                     3       2  ...   
19458                        0                     3       2  ...   
19459                        0                     3       2  ...   
...                        ...                   ...     ...  ...   
39163                        1                     2       2  ...   
39202                        2                     1       2  ...   
39203                        2                     1       2  ...   
39207                        2                     2       2  ...   
43562                        0                     1       2  ...   

       deposit_type  agent  company  days_in_waiting_list  customer_type  \
11355             0   15.0     -1.0                     2              3   
17026             0   -1.0     -1.0                     2              2   
19457             1   44.0     -1.0                     3              2   
19458             1   44.0     -1.0                     3              2   
19459             1   44.0     -1.0                     3              2   
...             ...    ...      ...                   ...            ...   
39163             0    3.0     -1.0                     4              3   
39202             0    3.0     -1.0                     4              3   
39203             0    3.0     -1.0                     4              3   
39207             0    3.0     -1.0                     4              3   
43562             1   -1.0     -1.0                     2              3   

         adr  required_car_parking_spaces  total_of_special_requests  persone  \
11355   0.00                            0                          0      2.0   
17026  48.00                            0                          0      2.0   
19457  75.00                            0                          0      2.0   
19458  75.00                            0                          0      2.0   
19459  75.00                            0                          0      2.0   
...      ...                          ...                        ...      ...   
3916

Controllare i valori in base al tipo di stanza prenotata e il tipo di stanza assegnata e come influisce il cambio nelle cancellazioni.

controllare la distribuzione in base al "customer Type". (fino a qui Salvo)

Controllare la distribuzione dei dati in base alla durata del soggiorno (fare una colonna "days" (weekend + week))

Controllare la distribuzione di dati in base al tipo di prenotazione e pasti prenotati (b&b etc.)

Controllare il numero di cancellazioni in base al deposito fatto oppure no (deposit_type)

Controllare la distribuzione di chi ha richiesto un parcheggio e chi no